client/core/core.go

```diff
@@ -4563,16 +4565,64 @@ func handleMatchProofMsg(c *Core, dc *dexConnection, msg *msgjson.Message) error
 		return fmt.Errorf("match proof note unmarshal error: %w", err)
 	}
 
-	// Expire the epoch
-	dc.setEpoch(note.MarketID, note.Epoch+1)
+	// Expire previous epoch.
+	dc.tryUpdateEpoch(note.MarketID, note.Epoch+1)
 
 	book := dc.bookie(note.MarketID)
 	if book == nil {
 		return fmt.Errorf("no order book found with market id %q",
 			note.MarketID)
 	}
 
-	return book.ValidateMatchProof(note)
+	err = book.ValidateMatchProof(note)
+	if err != nil {
+		return fmt.Errorf("match proof validation failed: %w", err)
+	}
+
+	err = validateMatchCSum(note.Preimages, note.CSum, dc.trackedTrades())
+	if err != nil {
+		// Validation can fail either due to server trying to cheat, or
+		// client loosing (e.g. due to restarting, since we don't persistently store these at the moment)
+		// his csums for the trades involved.
+		//
+		// We'll just warn the user for now (maybe we should handle this in a stricter way in the future).
+		c.log.Warnf("validate match checksum for epoch %d, err: %v", note.Epoch, err)
+	}
+
+	return nil
+}
+
+// validateMatchCSum for each preimage in the match validates
+// commited checksums on corresponding orders (trades)
+// against csum for this match, making sure they are of the same value.
+//
+// This is necessary to ensure that for an order O
+// "server preimage request" and "O being matched" happen in the same epoch,
+// otherwise server could game match shuffle, if it sees O's preimage
+// before committing to csum for a match.
+func validateMatchCSum(preimages []order.Preimage, matchCSum dex.Bytes, trades []*trackedTrade) error {
+	for _, preimage := range preimages {
+		err := validateMatchCSumForPreimage(preimage, matchCSum, trades)
```

Comment:
This is going to error virtually every time. The server is sending every preimage of every order from the epoch, not just our own.

Reply:
hmm, good point, haven't realized that `trackedTrade` represents an order issued by "this client only" and not just any order with additional meta,

shall I modify a description for `trackedTrade` ?

```
// trackedTrade is an order, its matches, and its cancel order, if applicable.
// The trackedTrade has methods for handling requests from the DEX to progress
// match negotiation.
```
to this
```
// trackedTrade is an order (issued by this client), its matches, and its cancel
// order, if applicable. The trackedTrade has methods for handling requests
// from the DEX to progress match negotiation.
```

as well as add a comment for `dexConnection.trades` to clarify that ?